frontends/concrete-python/concrete/fhe/extensions/truncate_bit_pattern.py

```diff
@@ -0,0 +1,249 @@
+"""
+Declaration of `truncate_bit_pattern` extension.
+"""
+
+import threading
+from copy import deepcopy
+from typing import Any, Callable, Dict, Iterable, List, Tuple, Union
+
+import numpy as np
+
+from ..dtypes import Integer
+from ..mlir.utils import MAXIMUM_TLU_BIT_WIDTH
+from ..representation import Node
+from ..tracing import Tracer
+from ..values import ValueDescription
+
+local = threading.local()
+
+# pylint: disable=protected-access
+local._is_adjusting = False
+# pylint: enable=protected-access
+
+
+class Adjusting(BaseException):
+    """
+    Adjusting class, to be used as early stop signal during adjustment.
+    """
+
+    truncator: "AutoTruncator"
+    input_min: int
+    input_max: int
+
+    def __init__(self, truncator: "AutoTruncator", input_min: int, input_max: int):
+        super().__init__()
+        self.truncator = truncator
+        self.input_min = input_min
+        self.input_max = input_max
+
+
+class AutoTruncator:
+    """
+    AutoTruncator class, to optimize for the number of msbs to keep during truncate operation.
+    """
+
+    target_msbs: int
+
+    is_adjusted: bool
+    input_min: int
+    input_max: int
+    input_bit_width: int
+    lsbs_to_remove: int
+
+    def __init__(self, target_msbs: int = MAXIMUM_TLU_BIT_WIDTH):
+        # pylint: disable=protected-access
+        if local._is_adjusting:
+            message = (
+                "AutoTruncators cannot be constructed during adjustment, "
+                "please construct AutoTruncators outside the function and reference it"
+            )
+            raise RuntimeError(message)
+        # pylint: enable=protected-access
+
+        self.target_msbs = target_msbs
+
+        self.is_adjusted = False
+        self.input_min = 0
+        self.input_max = 0
+        self.input_bit_width = 0
+        self.lsbs_to_remove = 0
+
+    @staticmethod
+    def adjust(function: Callable, inputset: Union[Iterable[Any], Iterable[Tuple[Any, ...]]]):
+        """
+        Adjust AutoTruncators in a function using an inputset.
+        """
+
+        # pylint: disable=protected-access,too-many-branches
+
+        try:  # extract underlying function for decorators
+            function = function.function  # type: ignore
+            assert callable(function)
+        except AttributeError:
+            pass
+
+        if local._is_adjusting:
+            message = "AutoTruncators cannot be adjusted recursively"
+            raise RuntimeError(message)
+
+        try:
+            local._is_adjusting = True
+            while True:
```

Comment:
this is difficult to understand, a comment could help
```
# processing each truncator one by one
while True:
```

Reply:
Or maybe
```
empty_input_set = True
could_have_another_truncator = True
while could_have_another_truncator:
     for ...:
           empty_input_set = False
           try:
           ...
           except:
           ....
           else:
               assert truncator is None
               could_have_another_truncator = False
               break # the for loop
     if  empty_input_set and could_have_another_truncator:
          message = "AutoTruncators cannot be adjusted with an empty inputset"
          raise ValueError(message)     
```

Reply:
`could_have_another_truncator` in this code is pretty much useless, it's only set to false before breaking the loop manually, so it's never read as False.

Also, I don't agree. When I see `while True:`, I think `This is a dynamic loop with a complex condition which is hard to express, there will be a "break" in the body` and I keep reading instead of thinking `This loop needs to have a condition no matter how complex it is to express`.

I agree that it can use some comments though, so I'll add them :+1:

Reply:
Done.